axlearn/common/input_grain_lm.py

```
@@ -23,12 +24,303 @@ def __call__(self, ids: Tensor, *, max_len: int) -> Tensor:
+    """
+
+    def __init__(
+        self,
+        parent: grain.DatasetIterator,
+        *,
+        max_len: int,
+        window_size: Optional[int] = None,
+        input_key: str = "target_labels",
+    ):
+        super().__init__(parent)
+        self._max_len = max_len
+        self._window_size = window_size
+        self._input_key = input_key
+
+        # Index of the parent.
+        self._index = 0
+        self._current_token_count = 0
+        self._current_examples_list = []
+        # Both of them are used for checkpointing.
+        self._parent_sequence_start_state = None
+        self._parent_sequence_end_state = None
+        self._parent_state = self._parent.get_state()
+
+    def _reach_window_limit(self) -> bool:
+        """Determines if we have already reached window limit."""
+        return self._window_size is not None and self._index % self._window_size == 0
+
+    def _pop_element(self) -> Optional[dict]:
+        """Pops element from self._current_example_list."""
+        if not self._current_examples_list:
+            return None
+        concat_target_labels = np.concatenate(
+            [x[self._input_key] for x in self._current_examples_list], axis=-1
+        )
+        # Total tokens to pop could be up to self._max_len
+        total_tokens_to_pop = min(len(concat_target_labels), self._max_len)
+        self._current_token_count -= total_tokens_to_pop
+        if self._current_token_count > 0:
+            self._current_examples_list = [{self._input_key: concat_target_labels[self._max_len :]}]
+            self._parent_sequence_start_state = self._parent_sequence_end_state
+        else:
+            self._current_examples_list = []
+            self._parent_sequence_start_state = None
+        assert self._current_token_count >= 0
+
+        # If all the concat target labels is empty, early return.
+        if total_tokens_to_pop == 0:
+            return None
```

Maybe we can raise StopIteration instead of returning None?

---

Thanks for the suggestion. We have already handled StopIteration in the __next__ function. This return condition is used in some corner condition, especially when some examples are broken and has 0 length:

Let's say we have a window_size=3, sequence length of 8, and three examples [3, 5, 0]: We will first genreate the first sequence with the first two examples.

When we consume the second example, we reach the window limit, but the entire sequence is still empty.